test/Kestrel.Core.Tests/Http1ConnectionTests.cs

```diff
@@ -792,6 +793,34 @@ public void ThrowsWhenMaxRequestBodySizeIsSetToANegativeValue()
             Assert.StartsWith(CoreStrings.NonNegativeNumberOrNullRequired, ex.Message);
         }
 
+        [Fact]
+        public async Task ConsumesRequestBeforeClosingConnection()
+        {
+            var httpApplication = new DummyApplication(async context =>
+            {
+                var buffer = new byte[10];
+                await context.Response.Body.WriteAsync(buffer, 0, 10);
+            });
+
+            var requestProcessingTask = _http1Connection.ProcessRequestsAsync(httpApplication);
+
+            var requestCompleted = false;
+            _application.Output.OnReaderCompleted((exception, state) => { requestCompleted = true; }, null);
+
+            var data = Encoding.ASCII.GetBytes("POST / HTTP/1.1\r\nHost:\r\nConnection: close\r\ncontent-length: 1\r\n\r\n");
+            await _application.Output.WriteAsync(data);
+
+            // Wait for ProcessRequestsAsync to process the request
+            // Unfortunately there is no way to know when it is waiting on the request body
+            await Task.Delay(150);
```

Comment:
Get rid of the delay and instead use a ManualResetEventSlim and wait for that with a (very long) timeout. E.g:

```c#
var requestCompletedMre = new ManualResetEventSlim();

// ...

_application.Output.OnWriterCompleted((exception, state) => { requestCompletedMre.Set() }, null);

// ...

Assert.True(requestCompletedMre.Wait(TestConstants.DefaultTimeout);
```

Reply:
Ignore the previous comment. I misread the test. I see you're verifying that OnReaderCompleted *isn't* called. The problem I see is that unless you wire up something to consume the application output, there will never be a reader to get completed, so OnReaderCompleted can't be called.

Reply:
Sorry it's not really clear this test.
The idea is that, if it is not waiting for the body [here](https://github.com/aspnet/KestrelHttpServer/pull/2314/files/de2ecaa53051e669fc4ba9309b9646c30320b826#diff-71145e700087089041944298eab15d6cR539) it will carry on and complete the pipe [here](https://github.com/aspnet/KestrelHttpServer/pull/2314/files/de2ecaa53051e669fc4ba9309b9646c30320b826#diff-71145e700087089041944298eab15d6cR603) given it is not a connection to be kept alive.

In the end this is the same as checking if the requestProcessingTask has complete, but I thought it was better to check the pipe given that the test is about making sure it doesn't close the connection before consuming the body. The task termination feels like an implementation detail.

This is both testing that the request is not completed before receiving the body and that it is completed afterwards, otherwise it would timeout on waiting the requestProcessingTask. Probably for symmetry it can be tested `Assert.True(requestCompleted);` but testing that the connection is closed after it's not really the responsibility of this test.

I don't like this test at all and I'm happy to scrap it.

I can have another go tomorrow to see if I can mock the pipe to be notified when the processing task is waiting for data. Or any other suggestion is welcome.

Reply:
This test is making more sense to me now. I have to admit the transport/application pipe names always mess me up, so that was part of the confusion.

I wonder if you would be better off creating a custom mock Http1Connection that overrides CreateMessageBody() to return a mock MessageBody.

At the end of the test after the requestProcessingTask completes, you can then verify that ConsumeAsync did not got called on the mock message body. This way we don't need to introduce a new arbitrary timeouts that always seem to cause problems no matter how conservative we make them.

Thanks!

Reply:
Good call! I didn't notice `CreateMessageBody` was virtual.
I haven't created a separate test object or injected a `MessageBody` factory in the test object as it seems overkill.

I have renamed the test as it is only verified that the body get consumed, not that it actually happens before closing the connection but I believe this is enough and it makes the test a lot more clean.